security/advancedtls/provider_test.go

```diff
@@ -0,0 +1,261 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	"google.golang.org/grpc/security/advancedtls/testdata"
+)
+
+// The IdentityPemFileProvider is tested in different stages.
+// At stage 0, we create an IdentityPemFileProvider with empty initial files.
+// The KeyMaterial will be empty since the initial file contents are empty.
+// At stage 1, we copy the first set of certFile and keyFile to the temp files
+// that are watched by the goroutine.
+// The KeyMaterial is expected to be updated if there is a matching key-cert pair.
+// At stage 2, we copy the second set of certFile and keyFile to the temp files
+// and verify the credential files are updated.
+// The KeyMaterial is expected to be updated if there is a matching key-cert pair.
+// At stage 3, we clear the file contents of temp files.
```

Comment:
Can we also add a stage 4 that when the key and cert mismatch(like you use client_key_1.pem in combination withclient_cert_2.pem), and expect the update should be skipped?

Reply:
Actually I added that in the third test case.

Reply:
Isn't in stage 3 we just clear the file contents, instead of causing a mismatch? The mismatch issue is a real issue that could happen when the file rotation for cert is successful, while for the key failed. I think it could be good if we explicitly test this scenario(although from the reading's perspective, it is the same as reading empty files).

Reply:
Yup, I mean in the third test case, we cover this mismatch case in the second pair. The code is [here.](https://github.com/grpc/grpc-go/blob/add4d6eefab63d85f262b7f4b58eae4a28eb4387/security/advancedtls/provider_test.go#L92-L99)

Reply:
Oh ok. Sorry I was reading it wrong. It looks good then!